etw/rust/tracelogging/src/native.rs

```diff
@@ -272,19 +300,39 @@ impl ProviderContextInner {
     fn unregister(&mut self) -> u32 {
         let result;
 
-        let was_busy = self.busy.swap(true, atomic::Ordering::Acquire);
-        if was_busy {
-            result = 0;
-        } else {
-            if self.reg_handle == 0 {
+        #[cfg(all(windows, feature = "etw", feature = "user_mode"))]
+        {
+            let was_busy = self.busy.swap(true, atomic::Ordering::Acquire);
+            if was_busy {
                 result = 0;
             } else {
-                result = unsafe { EventUnregister(self.reg_handle) };
-                self.level = -1;
-                self.reg_handle = 0;
+                if self.reg_handle == 0 {
+                    result = 0;
+                } else {
+                    result = unsafe { EventUnregister(self.reg_handle) };
+                    self.level = -1;
+                    self.reg_handle = 0;
+                }
+
+                self.busy.swap(false, atomic::Ordering::Release);
+            }
+        }
+        #[cfg(all(windows, feature = "etw", feature = "kernel_mode"))]
+        {
+            let was_busy = self.busy.swap(true, atomic::Ordering::Acquire);
+            if was_busy {
+                result = 0;
+            } else {
+                if self.reg_handle == 0 {
+                    result = 0;
+                } else {
+                    result = unsafe { EtwUnregister(self.reg_handle) };
```

Comment:
Why did this duplicate a large chunk of code instead of just one line?

Reply:
Fixed